concrete-core-bench/src/benchmark.rs

```diff
@@ -0,0 +1,144 @@
+        criterion: &mut Criterion,
+        batch_size: Option<u64>,
+    ) {
+        let mut group = criterion.benchmark_group(format!(
+            "{}<{}, {}, {}>",
+            type_name::<Self>(),
+            type_name::<Precision>(),
+            type_name::<Engine>(),
+            type_name::<RelatedEntities>()
+        ));
+        for params in Self::generate_parameters_iterator() {
+            Self::bench(maker, engine, params, &mut group, batch_size);
+        }
+        group.finish();
+    }
+
+    /// Benchmark one parameter in the fixture benchmark group.
+    fn bench(
+        maker: &mut Maker,
+        engine: &mut Engine,
+        parameters: Self::Parameters,
+        fixture_group: &mut BenchmarkGroup<WallTime>,
+        batch_size: Option<u64>,
+    ) {
+        let batch_size = batch_size.unwrap_or(DEFAULT_BATCH_SIZE);
+
+        // We generate the prototypes once (used for all repetitions of the benchmark).
+        let repetition_proto = Self::generate_random_repetition_prototypes(&parameters, maker);
+        let sample_proto =
+            Self::generate_random_sample_prototypes(&parameters, maker, &repetition_proto);
+
+        // The benchmark itself does not use the standard iterators of `criterion`, because the
+        // pre-context is passed by value, and the post-context must be gathered to be
+        // properly disposed afterward.
+        fixture_group.bench_with_input(
+            BenchmarkId::from_parameter(format!("{:?}", parameters)),
+            &(repetition_proto, sample_proto),
+            |b, (repetition_proto, sample_proto)| {
+                b.iter_custom(|iterations_to_go| {
+                    // We initialize the vectors which will store the input and output values of the
+                    // batch.
+                    let mut inputs_vec: Vec<Self::PreExecutionContext> =
+                        Vec::with_capacity(batch_size as usize);
+                    let mut outputs_vec: Vec<Self::PostExecutionContext> =
+                        Vec::with_capacity(batch_size as usize);
+
+                    // We iterate until we reached the number of iterations to be performed.
+                    let mut iterations_counter = 0;
+                    let mut duration = Duration::new(0, 0);
+                    while iterations_counter < iterations_to_go {
+                        let batch_size = min(batch_size, iterations_to_go - iterations_counter);
+                        iterations_counter += batch_size;
+
+                        (0..batch_size).for_each(|_| {
+                            inputs_vec.push(Self::prepare_context(
+                                &parameters,
+                                maker,
+                                repetition_proto,
+                                sample_proto,
+                            ))
+                        });
+
+                        let start = Instant::now();
+                        inputs_vec.drain(..).for_each(|input| {
```

Comment:
Here I guess even though you put drain(..) it's going to clear only 1 input at each iteration?

Reply:
What `drain` does is that it creates an iterator that goes over all the members of `input_vec` and leave the vec empty (as if `clear` had been called). I feel like this is what we want here. Does that make sense to you ?

Reply:
Ah ok, yes it makes sense. I'm not sure we absolutely need it (we could just let the input/output vecs be dropped when the bench is over no?).

Reply:
-> yes we need it to reuse the same input/output vecs for all the batches, it's ok for me.